#!/applications/R/R-3.4.0/bin/Rscript

### Perform differential expression analysis using read counts generated by Salmon

# R version 3.4.0
# DESeq2 version 1.16.1
# Note that this R version or later is required for DESeq2 version 1.16 or later,
# in which "the log2 fold change shrinkage is no longer default for the DESeq and nbinomWaldTest functions".
# DESeq2 version 1.16 introduces "a separate function lfcShrink, which performs log2 fold change shrinkage
# for visualization and ranking of genes." (see https://support.bioconductor.org/p/95695/ and
# https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#changes)

# Usage:
# ./DESeq2.R 0.01 wt hta6

args <- commandArgs(trailingOnly = T)
#FDRnum <- 0.01
#FDRchar <- "0.01"
#genotype1 <- "wt"
#genotype2 <- "hta6"
FDRnum <- as.numeric(args[1])
FDRchar <- as.character(args[1])
genotype1 <- as.character(args[2])
genotype2 <- as.character(args[3])

library(DESeq2)
print(packageVersion("DESeq2"))
#[1] ‘1.16.1’

inDir <- "/home/ajt200/analysis/180928_Pallas_RNAseq_series1/fastq_pooled/trimmed_HEADCROP12_CROP60/DESeq2_genes_fakeReps/"
outDir <- paste0(inDir, "FDR", FDRchar, "/")
plotDir <- paste0(outDir, "plots/")
system(paste0("[ -d ", outDir, " ] || mkdir ", outDir))
system(paste0("[ -d ", plotDir, " ] || mkdir ", plotDir))

# Load gene-level counts as R list object named "txi"
load(paste0(inDir, "tximport.RData"))

sampleTable <- data.frame(sample = c("wt RNA-seq Rep1",
                                     "hta6 RNA-seq Rep1",
                                     "hta7 RNA-seq Rep1",
                                     "cmt3 RNA-seq Rep1",
                                     "hta6 hta7 RNA-seq Rep1",
                                     "cmt3 hta7 RNA-seq Rep1",
                                     "cmt3 hta6 hta7 RNA-seq Rep1"),
                          condition = factor(rep.int(c("wt", "hta6", "hta7", "cmt3",
                                                       "hta6 hta7", "cmt3 hta7", "cmt3_hta6_hta7"),
                                                     times = c(1, 1, 1, 1, 1, 1, 1))))
rownames(sampleTable) <- colnames(txi$counts)
print(sampleTable)
#                              sample      condition

## Differential expression analysis

# Re-import DESeqDataSet to redefine "dds"
dds <- DESeqDataSetFromTximport(txi = txi,
                                colData = sampleTable,
                                design = ~condition)

# Pre-filter the datasets
print(nrow(dds))
#[1] 27586 
# Retain only rows that have more than a single count across all samples
dds <- dds[rowSums(counts(dds)) > 1,]
print(nrow(dds))
#[1] 

# Contrast: genotype2 vs genotype1
dds_geneEstDispersions <- estimateDispersionsGeneEst(dds)
dispersions(dds_geneEstDispersions) <- mcols(dds_geneEstDispersions)$dispGeneEst
res_geno2Vgeno1 <- results(dds_geneEstDispersions,
                           test = nbinomWaldTest,
                           contrast = c("condition", genotype2, genotype1),
                           alpha = FDRnum)
print(mcols(res_kssVwt, use.names = T))
#DataFrame with 6 rows and 2 columns
#                       type                                 description
#                <character>                                 <character>
#baseMean       intermediate   mean of normalized counts for all samples
#log2FoldChange      results log2 fold change (MLE): condition kss vs wt
#lfcSE               results         standard error: condition kss vs wt
#stat                results         Wald statistic: condition kss vs wt
#pvalue              results      Wald test p-value: condition kss vs wt
#padj                results                        BH adjusted p-values

print(summary(res_kssVwt))

print(table(res_kssVwt$padj < FDRnum))

print(sum(!is.na(res_kssVwt$padj)))

####
# MA-plots
# These show the log2 fold changes in gene expression values for a variable (e.g., treated vs untreated)
# over a the mean of normalized counts for all the samples in the DESeqDataSet

# DESeq2 provides for use of a "Bayesian procedure to moderate (or "shrink") log2 fold changes from genes
# with very low counts and highly variable counts"
# Before generating an MA-plot, the lfcShrink() function should be used to moderate the log2(fold changes) for the contrast
res_kssVwt_lfcShrink <- lfcShrink(dds_kssVwt,
                                  contrast = c("condition", "kss", "wt"),
                                  res = res_kssVwt)
pdf(paste0(plotDir, "MAplot_res_kssVwt_", FDRchar, "_lfcShrink_genes.pdf"), height = 5, width = 6)
par(mfrow = c(1, 1))
par(mar = c(4.1, 4.1, 2.1, 2.1))
plotMA(res_kssVwt_lfcShrink,
       ylim = c(min(res_kssVwt_lfcShrink$log2FoldChange),
                max(res_kssVwt_lfcShrink$log2FoldChange)),
       xlab = "Mean normalized expression",
       ylab = expression(Log[2]~fold~change),
       colSig = "dodgerblue3",
       colNonSig = "black",
       main = bquote(italic("kyp suvh5 suvh6")*" vs wild type (FDR < "*.(FDRchar)*")"))
#abline(h = c(1, -1), lwd = 1.5, lty = 2, col = "red")
dev.off()


## Exporting results

# Subset results table to significant (padj < FDRnum) down-regulated genes
res_kssVwt_lfcShrink_Chr <- res_kssVwt_lfcShrink[!grepl("ATM", rownames(res_kssVwt_lfcShrink)) &
                                                 !grepl("ATC", rownames(res_kssVwt_lfcShrink)), ]
res_kssVwt_lfcShrink_Chr_downReg <- res_kssVwt_lfcShrink_Chr[!is.na(res_kssVwt_lfcShrink_Chr$padj) &
                                                             res_kssVwt_lfcShrink_Chr$padj < FDRnum &  
                                                             res_kssVwt_lfcShrink_Chr$log2FoldChange < 0, ]
# OR
#res_kssVwt_lfcShrink_Chr_downReg2 <- subset(res_kssVwt_lfcShrink_Chr, padj < FDRnum & log2FoldChange < 0)

# Sort by increasing log2 fold change estimate
# (genes with strongest down-regulation)
res_kssVwt_lfcShrink_Chr_downRegSorted <- res_kssVwt_lfcShrink_Chr_downReg[order(res_kssVwt_lfcShrink_Chr_downReg$log2FoldChange), ]
 

# Subset results table to significant (padj < FDRnum) up-regulated genes
res_kssVwt_lfcShrink_Chr_upReg <- res_kssVwt_lfcShrink_Chr[!is.na(res_kssVwt_lfcShrink_Chr$padj) &
                                                           res_kssVwt_lfcShrink_Chr$padj < FDRnum &  
                                                           res_kssVwt_lfcShrink_Chr$log2FoldChange > 0, ]
# OR
#res_kssVwt_lfcShrink_Chr_upReg2 <- subset(res_kssVwt_lfcShrink_Chr, padj < FDRnum & log2FoldChange > 0)

# Sort by decreasing log2 fold change estimate
# (genes with strongest up-regulation)
res_kssVwt_lfcShrink_Chr_upRegSorted <- res_kssVwt_lfcShrink_Chr_upReg[order(res_kssVwt_lfcShrink_Chr_upReg$log2FoldChange,
                                                                             decreasing = T), ]

# Convert DataFrame objects (IRanges package) to data.frame objects that can be
# processed by write.table
res_kssVwt_lfcShrink_Chr_downRegSortedDF <- as.data.frame(res_kssVwt_lfcShrink_Chr_downRegSorted)
res_kssVwt_lfcShrink_Chr_upRegSortedDF <- as.data.frame(res_kssVwt_lfcShrink_Chr_upRegSorted)
write.table(res_kssVwt_lfcShrink_Chr_downRegSortedDF,
            file = paste0(outDir, "res_kssVwt_", FDRchar, "_lfcShrink_Chr_downRegSortedDF_genes.txt"))
write.table(res_kssVwt_lfcShrink_Chr_upRegSortedDF,
            file = paste0(outDir, "res_kssVwt_", FDRchar, "_lfcShrink_Chr_upRegSortedDF_genes.txt"))

res_kssVwt_lfcShrink_Chr_downReg_geneIDs <- rownames(res_kssVwt_lfcShrink_Chr_downRegSortedDF)
res_kssVwt_lfcShrink_Chr_upReg_geneIDs <- rownames(res_kssVwt_lfcShrink_Chr_upRegSortedDF)
write.table(res_kssVwt_lfcShrink_Chr_downReg_geneIDs,
            file = paste0(outDir, "res_kssVwt_", FDRchar, "_lfcShrink_Chr_downReg_geneIDs.txt"))
write.table(res_kssVwt_lfcShrink_Chr_upReg_geneIDs,
            file = paste0(outDir, "res_kssVwt_", FDRchar, "_lfcShrink_Chr_upReg_geneIDs.txt"))


## Generate plots of log-transformed normalized counts at differentially expressed genes

downReg_plotDir <- paste0(plotDir, "downReg_counts/")
upReg_plotDir <- paste0(plotDir, "upReg_counts/")
system(paste0("[ -d ", downReg_plotDir, " ] || mkdir ", downReg_plotDir))
system(paste0("[ -d ", upReg_plotDir, " ] || mkdir ", upReg_plotDir))

library(ggplot2)
library(ggbeeswarm)
library(parallel)
library(org.At.tair.db)
TAIRsymbol <- org.At.tairSYMBOL
# Get TAIR gene identifiers that are mapped to a gene symbol
mapped_geneIDs <- mappedkeys(TAIRsymbol)

dds <- DESeqDataSetFromTximport(txi = txi,
                                colData = sampleTable,
                                design = ~condition)

res_kssVwt_lfcShrink_Chr_downReg_geneIDs <- as.character(read.table(file = paste0(outDir,
                                                                               "res_kssVwt_", FDRchar, "_lfcShrink_Chr_downReg_geneIDs.txt"))$x)
res_kssVwt_lfcShrink_Chr_upReg_geneIDs <- as.character(read.table(file = paste0(outDir,
                                                                             "res_kssVwt_", FDRchar, "_lfcShrink_Chr_upReg_geneIDs.txt"))$x)

# Count-plotting function
countPlotFun <- function(geneID, contrast, plotDir) {
  geneCounts <- plotCounts(dds,
                           gene = geneID,
                           intgroup = c("condition", "sample"),
                           normalized = T,
                           transform = T,
                           returnData = T)
  geneCounts$condition <- factor(geneCounts$condition, levels = c("wt", "kss"))
  geneCountsPlot <- ggplot(geneCounts,
                           aes(x = condition,
                               y = count,
                               colour = sample)) +
                    scale_y_log10(breaks = pretty(geneCounts$count)) +
                    geom_beeswarm(cex = 3) +
                    xlab("Genotype") +
                    ylab("Log-transformed normalized count") +
                    labs(colour = "Sample") +
                    theme(plot.margin = grid::unit(c(0,0,0,0), "mm")) +
                    theme_classic() +
                    theme(panel.border = element_rect(colour = "black", fill = NA, size = 1)) +
                    ggtitle(paste0(geneID, " (", TAIRsymbol[[geneID]][1], ")")) +
                    theme(plot.title = element_text(hjust = 0.5))
  ggsave(geneCountsPlot,
         file = paste0(plotDir, contrast, "_", geneID, "_", gsub("/", "|", TAIRsymbol[[geneID]][1]), "_normalized_counts.pdf"),
         width = 16, height = 10, units = "cm")
}

mclapply(seq_along(res_kssVwt_lfcShrink_Chr_downReg_geneIDs), function(x) {
  countPlotFun(geneID = res_kssVwt_lfcShrink_Chr_downReg_geneIDs[x],
               contrast = paste0("kssVwt_", FDRchar, "_lfcShrink_Chr_downReg"),
               plotDir = downReg_plotDir)
}, mc.cores = 48)

mclapply(seq_along(res_kssVwt_lfcShrink_Chr_upReg_geneIDs), function(x) {
  countPlotFun(geneID = res_kssVwt_lfcShrink_Chr_upReg_geneIDs[x],
               contrast = paste0("kssVwt_", FDRchar, "_lfcShrink_Chr_upReg"),
               plotDir = upReg_plotDir)
}, mc.cores = 48)


